docs/proposals/templates.md

```diff
@@ -0,0 +1,145 @@
+  between dependent pieces
+* Support maintaining a library of templates within Kubernetes that can be accessed and instantiated by end users
+* Allow users to quickly and repeatedly deploy instances of well-defined application patterns produced by the community
+* Follow established Kubernetes API patterns by defining new template related APIs which consume+return first class Kubernetes 
+  API (and therefore json conformant) objects.
+
+We do not wish to invent a new turing complete templating language.  There are good options available 
+(e.g. https://github.com/mustache/mustache) for developers who want a completely flexible and powerful solution for creating 
+arbitrarily complex templates with parameters, and tooling can be built around such schemes.
+
+This desire for simplicity also intentionally excludes template composability/embedding as a supported use case.  
+
+Allowing templates to reference other templates presents versioning+consistency challenges along with making the template 
+no longer a self-contained portable object.  Scenarios necessitating multiple templates can be handled in one of several 
+alternate ways:
+
+* Explicitly constructing a new template that merges the existing templates (tooling can easily be constructed to perform this 
+  operation since the templates are first class api objects).
+* Manually instantiating each template and utilizing [service linking](https://github.com/kubernetes/kubernetes/pull/17543) to share 
+  any necessary configuration data.
+
+This document will also refrain from proposing server APIs or client implementations.  This has been a point of debate, and it makes
+more sense to focus on the template/parameter specification/syntax than to worry about the tooling that will process or manage the
+template objects.  However since there is a desire to at least be able to support a server side implementation, this proposal
+does assume the specification will be k8s API friendly.
+
+## Desired characteristics
+
+* Fully k8s object json-compliant syntax.  This allows server side apis that align with existing k8s apis to be constructed 
+  which consume templates and existing k8s tooling to work with them.  It also allows for api versioning/migration to be managed by
+  the existing k8s codec scheme rather than having to define/introduce a new syntax evolution mechanism.
+  * (Even if they are not part of the k8s core, it would still be good if a server side template processing+managing api supplied 
+    as an ApiGroup consumed the same k8s object schema as the peer k8s apis rather than introducing a new one)
+* Self-contained parameter definitions.  This allows a template to be a portable object which includes metadata that describe
+  the inputs it expects, making it easy to wrapper a user interface around the parameterization flow.
+* Support for non-string parameters.  K8S objects define fields of type string, int, bool, and byte[](base64).  The parameter
```

Comment:
My point with struct/map/list is that this is insufficient if ever we need to template out a whole struct, rather than a field or two.  I can buy that this is intentional, but nothing actually says that.

IntOrString, I guess it depends on how you intent to handle ints.

We also have some time fields - I guess those are strings to JSON.  In general, storing templatizations in-line means you need some way to know how to template for every single field.  What values are in-band and what are not?

That seems like lunacy to me.  I think the design overall would be more powerful and less delicate if it looked more like an ELF binary.  The body of the object is what it is, templated fields are just zero-values, but the whole thing is accompanied by a list of fixup definitions (field x.y[1].z needs the parameter named "GAMMA") and by a list of variable definitions (there's a parameter named "GAMMA" of type string, with description "I am human friendly").